dpcpp/preconditioner/batch_jacobi_kernels.hpp.inc

```diff
@@ -0,0 +1,254 @@
+    const int sg_tid = sg.get_local_id();
+    const int sg_global_id = item_ct1.get_global_linear_id() / sg_size;
+
+    // one subwarp per small diagonal block
+    const auto batch_idx = sg_global_id / num_blocks;
+    const auto block_idx = sg_global_id % num_blocks;
+
+    ValueType block_row[sg_size];
+    const auto block_size =
+        block_pointers[block_idx + 1] - block_pointers[block_idx];
+    assert(block_size <= sg_size);
+
+    const int* __restrict__ current_block_pattern =
+        blocks_pattern + gko::detail::batch_jacobi::get_block_offset(
+                             block_idx, blocks_cumulative_storage);
+    ValueType* __restrict__ current_block_data =
+        blocks +
+        gko::detail::batch_jacobi::get_global_block_offset(
+            batch_idx, num_blocks, block_idx, blocks_cumulative_storage);
+    const auto stride =
+        gko::detail::batch_jacobi::get_stride(block_idx, block_pointers);
+
+    // each thread of the subwarp stores the column of the dense block/row
+    // of the transposed block in its local memory
+    if (sg_tid < block_size) {
+        for (int a = 0; a < block_size; a++) {
+            const auto idx =
+                current_block_pattern[a * gko::detail::batch_jacobi::get_stride(
+                                              block_idx, block_pointers) +
+                                      sg_tid];  // coalesced accesses
+            ValueType val_to_fill = zero<ValueType>();
+            if (idx >= 0) {
+                assert(idx < nnz);
+                val_to_fill = A_vals[idx + nnz * batch_idx];
+            }
+            block_row[a] = val_to_fill;
+        }
+    }
+
+    int perm = -1;
+    // invert the transpose of the dense block.
+    // Note: Each thread of the subwarp has a row of the block
+    // to be inverted. (local id: 0 thread has 0th row, 1st
+    // thread has 1st row and so on..) If block_size <
+    // subwarp_size, then threads with local id >= block_size do
+    // not mean anything. Also, values in the block_row for
+    // index >= block_size are meaningless
+    invert_dense_block(block_size, block_row, perm, item_ct1);
+    sg.barrier();
+
+    // write back the transpose of the transposed inverse matrix to block
+    // array
+    for (int a = 0; a < block_size; a++) {
+        const int col_inv_transposed_mat = a;
+        const int col = sg.shuffle(perm, a);  // column permutation
+        const int row_inv_transposed_mat =
+            perm;  // accumulated row swaps during pivoting
+        const auto val_to_write = block_row[col];
+
+        const int row_diag_block = col_inv_transposed_mat;
+        const int col_diag_block = row_inv_transposed_mat;
+
+        if (sg_tid < block_size) {
+            current_block_data[row_diag_block * stride + col_diag_block] =
+                val_to_write;  // non-coalesced accesses due to pivoting
+        }
+    }
+}
+
+
+template <typename ValueType>
+__dpct_inline__ void transpose_block_jacobi_kernel(
```

Comment:
Maybe I missed it, but I do not find the corresponding function in cuda/hip

Reply:
Yes, you are right. This is actually not needed.